### ISSUE TITLE
small error in distance between ellipse code

### ISSUE THREAD

**Issue by wergil81** [mpewsey/mpewsey.github.io#5]
![image](https://github.com/mpewsey/mpewsey.github.io/assets/15847647/007cb8a6-a642-4d73-aa0a-c00f3c8f7c03)
 you forgot to add the ellipse center in the case of non-contained ellipses

     ...:     def find_distance(self, other, tolerance = 1e-8, max_iterations = 10000, learning_rate = 0.04, decay_rate = 0.002):
     ...:         """
     ...:         Finds the minimum distance to another ellipse using gradient descent.
     ...:         """
     ...:         r1 = self.rotation_matrix()
     ...:         r2 = other.rotation_matrix()
     ...:         dx0 = np.array([self.x - other.x, self.y - other.y])
     ...:         t1 = math.atan2(-dx0[1], -dx0[0]) - self.angle
     ...:         t2 = math.atan2(dx0[1], dx0[0]) - other.angle
     ...:         t = np.array([t1, t2])
     ...:         a1 = 0.5 * self.width
     ...:         b1 = 0.5 * self.height
     ...:         a2 = 0.5 * other.width
     ...:         b2 = 0.5 * other.height
     ...:         iterations = 0
     ...:         error = tolerance
     ...:         errors = []
     ...: 
     ...:         # If ellipse center is contained in another ellipse, use vertices to determine initial angles
     ...:         if self.contains([other.x, other.y]) or other.contains([self.x, self.y]):
     ...:             ts = np.linspace(0, 2 * math.pi, 24, endpoint = False)
     ...:             ts = np.array(np.meshgrid(ts, ts)).T.reshape(-1, 2)
     ...:             xe = a1 * np.cos(ts[:,0])
     ...:             ye = b1 * np.sin(ts[:,0])
     ...:             x1 = np.dot(np.column_stack([xe, ye]), r1.T) + [self.x, self.y]
     ...:             xe = a2 * np.cos(ts[:,1])
     ...:             ye = b2 * np.sin(ts[:,1])
     ...:             x2 = np.dot(np.column_stack([xe, ye]), r2.T) + [other.x, other.y]
     ...:             t = ts[np.argmin(np.linalg.norm(x1 - x2, axis = 1))]
     ...: 
     ...:         while error >= tolerance and iterations < max_iterations:
     ...:             cost = np.cos(t)
     ...:             sint = np.sin(t)
     ...:             x1 = np.dot(r1, [a1 * cost[0], b1 * sint[0]]) + **[self.x, self.y]**
     ...:             x2 = np.dot(r2, [a2 * cost[1], b2 * sint[1]]) + **[other.x, other.y]**
     ...:             dx1 = np.dot(r1, [-a1 * sint[0], b1 * cost[0]])
     ...:             dx2 = np.dot(r2, [-a2 * sint[1], b2 * cost[1]])
     ...:             delta = x1 - x2 + dx0
     ...:             dp = np.array([2 * np.dot(dx1, delta), -2 * np.dot(dx2, delta)])
     ...:             alpha = learning_rate * math.exp(-decay_rate * iterations)
     ...:             t -= alpha * dp
     ...:             error = np.max(np.abs(dp))
     ...:             errors.append(error)
     ...:             iterations += 1

**Comment by mpewsey** [comment]
The precomputed term `dx0` includes the ellipse centers, so there is no need to add them there.

**Comment by mpewsey** [comment]
Looking beyond the snippet you provided, it looks like computed distance `y = np.linalg.norm(x1 - x2)` is missing the `dx0` term. It should be `y = np.linalg.norm(x1 - x2 + dx0)`. I'll add that in.
